lib/makara/context.rb

```diff
@@ -1,83 +1,144 @@
+      # Called by `Proxy#stick_to_master!` to stick subsequent requests to
+      # master when using the given config
+      def stick(config_id, ttl)
+        current.stick(config_id, ttl)
       end
 
-      def set_current(context)
-        set(:makara_context_current,context)
+      def stuck?(config_id)
+        current.stuck?(config_id)
       end
 
-      def previously_stuck?(config_id)
-        previously_sticky.fetch(config_id) do
-          stuck?(Makara::Context.get_previous, config_id)
+      def commit(headers, cookie_options = {})
+        current.release_expired
+        if current.dirty?
+          cookie = DEFAULT_OPTIONS.merge(cookie_options)
+          Rack::Utils.set_cookie_header! headers, IDENTIFIER, cookie.merge(current.to_cookie_options)
         end
       end
 
-      # Called by `Proxy#stick_to_master!` to stick subsequent requests to
-      # master. They'll see the current context as their previous context
-      # when they're asking whether they should be stuck to master.
-      def stick(context, config_id, ttl)
-        Makara::Cache.write(cache_key_for(context, config_id), '1', ttl)
+      def release(config_id)
+        current.release(config_id)
       end
 
-      def stuck?(context, config_id)
-        !!Makara::Cache.read(cache_key_for(context, config_id))
+      def release_all
+        current.release_all
       end
 
       protected
-
-      def previously_sticky
-        fetch(:makara_previously_sticky) { Hash.new }
-      end
-
-      def cache_key_for(context, config_id)
-        "makara::#{context}-#{config_id}"
+      def current
+        fetch(:makara_current_context) { new({}) }
       end
 
       def fetch(key)
-        get(key) || set(key,yield)
+        get(key) || set(key, yield)
       end
 
       if Thread.current.respond_to?(:thread_variable_get)
         def get(key)
           Thread.current.thread_variable_get(key)
         end
 
-        def set(key,value)
-          Thread.current.thread_variable_set(key,value)
+        def set(key, value)
+          Thread.current.thread_variable_set(key, value)
         end
       else
         def get(key)
           Thread.current[key]
         end
 
-        def set(key,value)
+        def set(key, value)
           Thread.current[key]=value
         end
       end
 
+      private
+
+      # Pairs of {config_id}:{timestamp}, separated by "|"
+      # config_id1:1518270031.3132212|config_id2:1518270030.313232 ..
+      def parse(cookie_string)
+        return {} if cookie_string.empty?
+
+        states = cookie_string.split("|")
+        states.each_with_object({}) do |state, data|
+          config_id, timestamp = state.split(":")
+          data[config_id] = timestamp.to_f if config_id && timestamp
+        end
+      end
+    end
+
+    attr_accessor :data
+
+    def initialize(data)
+      @data = data
+      @dirty = false
+    end
+
+    def stick(config_id, ttl)
+      data[config_id] = Time.now.to_f + ttl.to_f
```

Comment:
I'm rethinking this based on the `write_to_cache` and "Full Request Sticking" comments below.
What is this method was: `def stick(config_id, ttl, write_to_cache=true)`

The "Full Request Sticking" concept says that i should stick the whole request even if the request is longer than the `ttl`. So we need to set something. Maybe `data` should be a hash like:
`data[config_id] = { expires_at: time_from_cookie, ttl: ttl_from_this_method, persist: true }`

`initialize` would set all the `expires_at` and this method would set the `ttl` if it's bigger than current (or unset) and set `persist` to `true` if `write_to_cache` was `true`.

Then below, we'd use this data to write the cookies. I think this `Time.now` stuff is dangerous because of Transactions in long requests. The TTL shouldn't really "kick in" until the request is over because that's when the data actually starts to replicate. So in `to_cookie_options` we would get the `Time.now` and then add all the `ttl` values if `persist` was true.

Reply:
In theory, you could also have a separate hash if that feels better and leave `data` as just what's in the cookie. Not sure if that is better or not. Seems like more to coordinate, but maybe there are advantages, too, when really going through it.

Reply:
Just thinking this through one more time as I walked away. Usually when I make one of these hashes with several keys that have various meanings (often called a data clump), I don't regret making an Object or Struct to represent it. Might be worth thinking about.